src/ssm-mem.c

```diff
@@ -54,30 +54,56 @@ struct ssm_object *ssm_new(uint8_t val_count, uint8_t tag) {
 void ssm_dup(struct ssm_mm *mm) { ++mm->ref_count; }
 
 void ssm_drop(struct ssm_mm *mm) {
-  if (--mm->ref_count == 0)
+  if (--mm->ref_count == 0) {
+    if (!ssm_mm_is_builtin(mm)) {
+      struct ssm_object *obj = container_of(mm, struct ssm_object, mm);
+      for (int i = 0; i < mm->val_count; i++) {
+        if (!ssm_mm_is_builtin(obj->payload[i].heap_ptr)) {
+          ssm_drop(obj->payload[i].heap_ptr);
+        }
+      }
+    }
     ssm_mem_free(mm, ssm_mm_is_builtin(mm) ? SSM_BUILTIN_SIZE(mm->tag)
                                            : SSM_OBJ_SIZE(mm->val_count));
+  }
 }
 
 struct ssm_mm *ssm_reuse(struct ssm_mm *mm) {
```

Comment:
This implementation does not seem complete: it only allows objects of the same type + variant to be reused, without the caller manually modifying the tag.

This API also differs from what is presented in the Perceus paper, where `new` takes an additional "memory token" parameter that `reuse` returns. This allows memory to be dropped more "eagerly", so that we do this:

```rust
let big_obj = ...;
let big_tok = reuse(big_obj);
///
let new_obj = new(vals, tag, big_tok);
```
The advantage here is that if `big_obj` contains many children, they will be dropped as soon as possible, at `reuse()`, so that the memory can be reclaimed in the code area highlighted by `///`. The semantics you seem to have implemented does not allow `reuse` and `new` to be separated, potentially missing out on this opportunity.

The downside of the API presented by Perceus is that it requires codegen to manage these tokens in addition to regular object pointers, complicating things somewhat, but I don't think that it's too much to handle. It's also somewhat unclear to me how big of a memory usage advantage their design will confer, because it really depends on how deeply nested objects are.

I think we should punt on implementing reuse until after we've gotten regular `new`, `dup`, and `drop` working completely, i.e., generating code that uses that too.